include/networkit/distance/STSP.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * STSP.hpp
+ *
+ *  Created on: 14.06.2019
+ *      Author: Eugenio Angriman <angrimae@hu-berlin.de>
+ */
+
+#ifndef STSP_H
+#define STSP_H
+
+#include <networkit/Globals.hpp>
+#include <networkit/base/Algorithm.hpp>
+#include <networkit/graph/Graph.hpp>
+
+namespace NetworKit {
+
+/**
+ * @ingroup distance
+ * Abstract base class for source-target shortest path algorithms.
+ */
+class STSP : public Algorithm {
+
+public:
+    /**
+     * Creates the STSP class for a graph @a G, source node @a source, and
+     * target node @a target.
+     *
+     * @param G The graph.
+     * @param source The source node.
+     * @param target The target node.
+     * @param storePred If true, the algorithm will also store the predecessors
+     * and reconstruct a shortest path from @a source and @a target.
+     */
+    STSP(const Graph &G, node source, node target, bool storePred = true)
+        : G(G), source(source), target(target), storePred(storePred) {
+        if (!G.hasNode(source))
+            throw std::runtime_error("Error: source node not in the graph!");
+        if (!G.hasNode(target))
+            throw std::runtime_error("Error: target node not in the graph!");
+        if (source == target)
+            INFO("Source and target nodes are equal!");
+    }
+
+    /**
+     * Returns a shortest path from the source node to the target node (without
+     * including them). Note: the shortest path can be constructed only if the
+     * algorithm is executed with @a storePred set to true.
+     *
+     * @return A shortest path from the @a source to @a target.
+     */
+    std::vector<node> getPath() const {
+        checkStoreDistances();
+        assureFinished();
+        return path;
+    }
+
+    /**
+     * Returns the predecessor nodes from the @a target to the source. Note:
+     * predecessors are stored only if the algorithm is executed with @a
+     * storePred set to true.
+     *
+     * @return The list of predecessors from @a target to @a source.
+     */
+    std::vector<node> getPredecessors() const {
+        checkStoreDistances();
+        assureFinished();
+        return pred;
+    }
+
+protected:
+    const Graph &G;
+    node source, target;
+    const bool storePred;
+    std::vector<uint8_t> visited;
+    uint8_t ts = 0;
+    static constexpr uint8_t ballMask = uint8_t(1) << 7;
```

Comment:
It is necessary to have `visited` in this class? This makes it impossible to use any implementation strategy other than bidirectional exploration.

Reply:
It is not, but then also `ts` and `ballMask` should be moved to the derived classes.
This choice was made to avoid to initialize and reset the `visited` vector two times, but if we want a more general implementation then I agree that all those members should be moved to the derived classes.

Reply:
Yes. Why does moving the vector to the derived classes imply that it needs to be initialized multiple times?

Reply:
Every time that the `run()` method is called we need to increase `ts` and, eventually, fill `visited` with zeros. Moving `visited` to the derived classes implies to rewrite the code that does this operation for each derived class if they use a bidirectional graph exploration as `BidirectionalBFS` and `BidirectionalDijkstra`. However, because we cannot assume that every derived class of `STSP` will use such strategy, I am fine to do so.